File: snowflake.lua
local core = require "snowflake.core"

local function gen62char()
    local map = {}
    for i=0,61 do
        local char
        if 10 <= i and i < 36 then
            char = string.char(65+i-10)
        elseif 36 <= i and i < 62 then
            char = string.char(97+i-36)
        else
            char = tostring(i)
        end
        map[i] = char
        map[char] = i
    end
    return map
end

local snowflake = {
	nodeid = nil,
	nodes = {},
    char62 = gen62char(),
}

function snowflake.init(nodeid,nodes)
	snowflake.nodeid = nodeid
	if nodes then
		for nodename,nodeid in pairs(nodes) do
			snowflake.node(nodename,nodeid)
		end
	end
end

function snowflake.node(nodename,nodeid)
	-- get
	if nodeid == nil then
		return snowflake.nodes[nodename]
	end
	-- register
	assert(snowflake.nodes[nodename] == nil)
	assert(snowflake.nodes[nodeid] == nil)
	snowflake.nodes[nodename] = nodeid
	snowflake.nodes[nodeid] = nodename
end


function snowflake.uuid(nodeid)
	if type(nodeid) == "string" then
		nodeid = snowflake.nodes[nodeid]
	end
	nodeid = nodeid or snowflake.nodeid
	return core.uuid(nodeid)
end

function snowflake._tostring(num,base)
    local list = {}
    while num > 0 do
        local divider = math.floor(num / base)
        local remainder = num % base
        num = divider
        table.insert(list,1,snowflake.char62[remainder])
    end
    return table.concat(list,"")
end

function snowflake.tostring(uuid,base)
    base = base or 10
    assert(base <= 62)
    local timestamp = snowflake.getTimestampByUUID(uuid)
    local nodeid = snowflake.getNodeIdByUUID(uuid)
    local sequence = snowflake.getSequenceByUUID(uuid)
    local list = {snowflake._tostring(timestamp,base),snowflake._tostring(nodeid,base),snowflake._tostring(sequence,base)}
    return table.concat(list,"-")
end

function snowflake._tointeger(str,base)
    local num = 0
    for i=1,#str do
        local b = string.sub(str,i,i)
        num = num * base + snowflake.char62[b]
    end
    return num
end

function snowflake.tointeger(str,base)
    base = base or 10
    assert(base <= 62)
    local timestamp,nodeid,sequence = string.match(str,"([0-9A-Za-z]+)-([0-9A-Za-z]+)-([0-9A-Za-z]+)")
    timestamp = snowflake._tointeger(timestamp,base)
    nodeid = snowflake._tointeger(nodeid,base)
    sequence = snowflake._tointeger(sequence,base)
    return snowflake.composeUUID(timestamp,nodeid,sequence)
end

function snowflake.getTimestampByUUID(uuid)
	return core.getTimestampByUUID(uuid)
end

function snowflake.getNodeIdByUUID(uuid)
	return core.getNodeIdByUUID(uuid)
end

function snowflake.getSequenceByUUID(uuid)
	return core.getSequenceByUUID(uuid)
end

function snowflake.getNodeNameByUUID(uuid)
	return snowflake.node(snowflake.getNodeIdByUUID(uuid))
end

function snowflake.composeUUID(timestamp,nodeid,sequence)
    return core.composeUUID(timestamp,nodeid,sequence)
end

return snowflake

File: snowflake.c
/*
 * 64位分布式id生成,由snowflake算法实现,构成如下
 *1bit  41bit     10bit     12bit
 *+--+---------+---------+--------+
 *|0 |timestamp|nodeId   |sequence|
 *+--+---------+---------+--------+
 */

#include <stdio.h>
#include <stdint.h>
#include <sys/time.h>
#include "snowflake.h"

#define TIMESTAMP_BIT (41)
#define NODEID_BIT (10)
#define SEQUENCE_BIT (12)
#define TIMESTAMP_MASK ((1ul<<TIMESTAMP_BIT)-1)
#define NODEID_MASK ((1<<NODEID_BIT)-1)
#define SEQUENCE_MASK ((1<<SEQUENCE_BIT)-1)

static uint64_t genTimestamp() {
    struct timeval tv;
    gettimeofday(&tv, 0);

    return (uint64_t)tv.tv_sec * 1000 + (uint64_t)tv.tv_usec / 1000;
}

static uint64_t genNodeId(int nodeId) {
    return (uint64_t) (nodeId % (1 << NODEID_BIT));
}

static uint64_t genIncrement(int* sequence) {
    *sequence = (*sequence + 1) & ~(1 << SEQUENCE_BIT);
    return *sequence;
}

uint64_t getTimestampByUUID(uint64_t uuid) {
    return (uuid >> (NODEID_BIT+SEQUENCE_BIT)) & TIMESTAMP_MASK;
}

int getNodeIdByUUID(uint64_t uuid) {
    return (uuid >> SEQUENCE_BIT) & NODEID_MASK;
}

int getSequenceByUUID(uint64_t uuid) {
    return uuid & SEQUENCE_MASK;
}

uint64_t composeUUID(uint64_t a,uint64_t b,uint64_t c) {
    return (a << (NODEID_BIT + SEQUENCE_BIT))
         | (b << SEQUENCE_BIT)
         | c;
}

uint64_t uuid(int nodeId,int* sequence) {
    uint64_t a = genTimestamp();
    uint64_t b = genNodeId(nodeId);
    uint64_t c = genIncrement(sequence);
    return composeUUID(a,b,c);
}
